File: CODIGOS3/codeabbey194.dart
import 'dart:math';

void main() {
  List<List<int>> points = [
    [113, 44],
    [87, 101],
    [64, 38],
    [52, 161],
    [108, 28],
    [27, 40],
    [145, 103],
    [161, 89],
    [96, 26],
    [83, 163],
    [69, 11],
    [15, 74],
    [50, 132],
    [48, 107],
    [134, 135],
    [37, 27],
    [3, 88],
  ];

  List<int> hull = _convexHull(points);

  print(hull);
}

List<int> _convexHull(List<List<int>> points) {
  int maxYIndex = points.indexWhere((point) => point[1] == points.map((point) => point[1]).reduce(max));
  List<int> hull = [maxYIndex];
  for (int i = 1; i <= points.length; i++) {
    while (hull.length > 2 &&
           cross(hull[hull.length - 2] - hull[hull.length - 1], points[i] - hull[hull.length - 1]) < 0) {
      hull.removeLast();
    }
    hull.add(i);
  }
  return hull;
}

int cross(List<int> a, List<int> b) {
  return a[0] * b[1] - a[1] * b[0];
}

File: CODIGOS3/codeabbey194.jl
function convex_hull(points)
    max_y_index = argmax(points .=> points[1][2])
    hull = [max_y_index]
    for i in 1:length(points)
      while length(hull) > 2 && cross(hull[end-1] - hull[end-2], points[i] - hull[end-1]) < 0
        hull = hull[1:end-1]
      end
      hull = push!(hull, i)
    end
    return hull
  end
  
  input = readlines(stdin)
  points = map(x -> parse.(Int, x), split(input[1], " "))
  hull = convex_hull(points)
  println(hull)
  